src/Dialect/Krnl/KrnlOps.td

```diff
@@ -26,10 +26,15 @@ def Krnl_Dialect : Dialect {
   let cppNamespace = "::mlir";
 }
 
+// Base class for Krnl dialect ops.
+class Krnl_Op<string mnemonic, list<OpTrait> traits = []> :
+    Op<Krnl_Dialect, mnemonic, !listconcat(traits,[MemRefsNormalizable])> {
+}
+
```

Comment:
Are we sure that all the memref are always normalizable in every of the the Krnl dialect operations?

Reply:
@AlexandreEichenberger Actually I don't know of any examples where memrefs cannot be normalized. I think you or Uday added `MemRefsNormalizable` trait in MLIR. Do you know what kinds of the example can't be normalized?

Reply:
I know I added it :-) but I don't recall precisely if everything can be normalized. I re-read the definition I wrote, quoted below.

----------------------

### MemRefsNormalizable

*   `OpTrait::MemRefsNormalizable` -- `MemRefsNormalizable`

This trait is used to flag operations that consume or produce values of `MemRef`
type where those references can be 'normalized'. In cases where an associated
`MemRef` has a non-identity memory-layout specification, such normalizable
operations can be modified so that the `MemRef` has an identity layout
specification. This can be implemented by associating the operation with its own
index expression that can express the equivalent of the memory-layout
specification of the MemRef type. See [the -normalize-memrefs pass].
(https://mlir.llvm.org/docs/Passes/#-normalize-memrefs-normalize-memrefs)

--------------------

So I think as long as the memory maps are ok, we should be fine. I was just surprised that we would assign this flag to all operations, but it might be ok. For example, what does it mean for a `krnl.define_loop` to be flagged as memref-normalizable when they don't have memrefs. But again, that might be ok.

Maybe someone else has an opinion on this, @tungld @doru1004 ?

Reply:
It looks like this way is still adhoc. What if we have ops not from krnl dialect, but some dialects in MLIR and they don't have MemRefNormalize trait too. It means every ops we use must have this trait, which is troublesome. 

For long-term, if memory-layout is simple (e.g. only tiling), we can directly generate normalized MemRef types for ops of interest only. Thus, don't need to normalize any MemRefs. It is perhaps easy if we use IndexExpr.

Reply:
Thanks for your comments.

I thought the traits was added in all ops in AffineDialect using base class, but I realized the base class was not for all ops. It was added only in  [AffineLoadOpBase](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Affine/IR/AffineOps.td#L482) and [AffineStoreOpBase](https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/Affine/IR/AffineOps.td#L841) and some other ops. 
So, would it be better to add only Krnl dialect ops whose input is memref as before?